### DISCUSSION TITLE
Some plug-in interface requests

### DISCUSSION THREAD

**Discussion by DreamMaoMao** [sxyazi/yazi#470] [4. Ideas]
hey, I want to implement a label jump plugin, and now I need these interfaces, are there any ready-made lua interfaces now?

- 1.Input parameter (void), get how many items are displayed in the current area, and be able to get the attributes of all the items(are displaying) like: (mime-type, path)
- 2.Input parameter (n), can get the type icon character of the nth item, like this: 
- 3.Input parameter (n, str), can modify the type icon character of the n th item to str (allow to modify to multiple characters, like this: AB)
- 4.Input parameter (n), can select the nth item with the cursor
- 5.Input parameter (fun), keyboard input event callback fun(event data), data contains the input character, and do not let other keyboard events be captured by other places.

@sxyazi

**Comment by sxyazi** [comment]
Thank you for summarizing this.

---

> 1.Input parameter (void), get how many items are displayed in the current area, and be able to get the attributes of all the items(are displaying) like: (mime-type, path)
> 
> 2.Input parameter (n), can get the type icon character of the nth item, like this: 

You can obtain the list of files in the current area using [`Folder:by_kind(Folder.CURRENT).window`](https://github.com/sxyazi/yazi/blob/cf769bf533259b2f157d813852327985bc0f55b1/yazi-plugin/preset/components/current.lua#L10).

It returns an array where each element is a file object. You can retrieve the file's icon using `file:icon()` and its mime-type using `file:mime()`, and so on. This is a list of all supported methods:

https://github.com/sxyazi/yazi/blob/cf769bf533259b2f157d813852327985bc0f55b1/yazi-plugin/src/bindings/file.rs#L15-L24

And:

https://github.com/sxyazi/yazi/blob/cf769bf533259b2f157d813852327985bc0f55b1/yazi-fm/src/lives/folder.rs#L47-L156

> 3.Input parameter (n, str), can modify the type icon character of the n th item to str (allow to modify to multiple characters, like this: AB)

https://github.com/sxyazi/yazi/blob/cf769bf533259b2f157d813852327985bc0f55b1/yazi-plugin/preset/components/folder.lua#L17

You can directly override the `Folder:icon(file)` method in your `init.lua` to do anything you wish, such as:

```lua
local old_icon_method = Folder.icon
function Folder:icon(file) 
    if my_plugin.some_key_pressed then
        return ui.Span("this is my custom icon")
    else
        return old_icon_method(self)
    end
end
```

> 4.Input parameter (n), can select the nth item with the cursor

We can add a dedicated command for it. Currently, you can only combine two commands, `arrow` and `select`, something like:

```lua
ya.manager_emit("arrow", { "5" })
ya.manager_emit("select", { state="true" })
```

It will select the fifth file.

> 5.Input parameter (fun), keyboard input event callback fun(event data), data contains the input character, and do not let other keyboard events be captured by other places.

We can also add it.

**Comment by DreamMaoMao** [comment]
First of all, thank you very much for your patient answer, I believe we can make yazi better together.

**Comment by DreamMaoMao** [comment]
I'll test the existing api first, and if you have the time, thank you very much for adding the last api.

**Comment by DreamMaoMao** [comment]
https://github.com/sxyazi/yazi/assets/30348075/46f4f214-e9b5-4319-a682-40367f77395d

hey,I think I have completed the testing of the previous interfaces. What should I do to bind my feature switch function in keymap.toml? just like triggering linemo.

**Comment by DreamMaoMao** [comment]
![image](https://github.com/sxyazi/yazi/assets/30348075/1892f4ff-4a28-49c0-bc86-6c2d104e6dab)

I can only manually change init.lua to statically start and shut down, I need a dynamic way to enable and shut down by key commands like other feature.
